### DISCUSSION TITLE
Rbatis V4 一些问题

### DISCUSSION THREAD

**Discussion by longzou** [rbatis/rbatis#258] [General]
主要还是一些常用的情形在V4中怎么实现：
1. 使用了crud!宏后，在执行insert、update操作时，如何指定表中某些字段不被更新的场景？
2. V4删除了Wrapper后，是不是动态查询只能通过html_sql、py_sql的方式来实现，以前可以使用Wrapper的r#if方式？特别是使用py_sql时，是不是会更容易因为“不小心”而产生错误。
3. 使用宏实现Update操作或Delete操作时，无法指定多个条件？？

**Comment by zhuxiujia** [comment]
> 主要还是一些常用的情形在V4中怎么实现：
> 
> 1. 使用了crud!宏后，在执行insert、update操作时，如何指定表中某些字段不被更新的场景？
> 2. V4删除了Wrapper后，是不是动态查询只能通过html_sql、py_sql的方式来实现，以前可以使用Wrapper的r#if方式？特别是使用py_sql时，是不是会更容易因为“不小心”而产生错误。
> 3. 使用宏实现Update操作或Delete操作时，无法指定多个条件？？

可以参考[abs-admin](https://github.com/rbatis/abs_admin)项目如何实现的.
insert,update 在字段Option为None的时候跳过，
update，delete可以选用 impl_update宏和impl_delete宏附加条件

**Comment by longzou** [comment]
impl_update宏，无法实现将某些字段update成为null值的操作。

**Comment by zhuxiujia** [comment]
> impl_update宏，无法实现将某些字段update成为null值的操作。

自定义使用py_sql 或者使用impl_select! 宏写语句更新对象

**Comment by longzou** [comment]
这个宏的实现中，也不太灵活，如下：
impl_select!(SysRole{select_list_by_ids(ids:&[String])=>
    "`where id in (`
     trim ',': for _,item in ids:
         #{item},
     )"});

这个时候就必须这个表名为sys_role，而无法用下面的方式来指定表名：
impl_select!(SysRole{select_list_by_ids(ids:&[String])=>
    "`where id in (`
     trim ',': for _,item in ids:
         #{item},
     )"}, "**tbl_sys_role**");

这种表名的映射还是需要的。

**Comment by zhuxiujia** [comment]
> 这个宏的实现中，也不太灵活，如下： impl_select!(SysRole{select_list_by_ids(ids:&[String])=> "`where id in (` trim ',': for _,item in ids: #{item}, )"});
> 
> 这个时候就必须这个表名为sys_role，而无法用下面的方式来指定表名： impl_select!(SysRole{select_list_by_ids(ids:&[String])=> "`where id in (` trim ',': for _,item in ids: #{item}, )"}, "**tbl_sys_role**");
> 
> 这种表名的映射还是需要的。

插入参数 table_name:&str即可，
看文档https://rbatis.github.io/rbatis.io/#/v4/?id=custom-table_name
